cross-compilation in openwrt fails

This came up during a test build for openwrt #62 but it possibly hints at something of wider significance.
Unfortunately, github is stupid and you can't even attach a txt file (seriously? Hello?!) so I will link the build log to a pastebin here that will expire in 90 days. http://paste.debian.net/164080/

The error depends on the way I build freelan and might be unrelated to freelan.  When I try to build the package by itself then boost fails to build.  It sounds like something is broken in openwrt here.  When building with "make menuconfig;make" I get the following error.

```
                 from build/release/include/freelan/configuration.hpp:55,
                 from build/release/apps/freelan/src/configuration_helper.hpp:50,
                 from build/release/apps/freelan/src/configuration_helper.cpp:47:
/media/rolf/DATA/tmp/compile/wrt/git/staging_dir/toolchain-mipsel_mips32_gcc-4.8-linaro_musl-1.1.10/include/sys/poll.h:1:2: error: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Werror=cpp]
 #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
  ^
cc1plus: all warnings being treated as errors
scons: *** [build/release/apps/freelan/src/configuration_helper.o] Error 1
scons: building terminated because of errors.
make[3]: *** [/media/rolf/DATA/tmp/compile/wrt/git/build_dir/target-mipsel_mips32_musl-1.1.10/freelan-2015-04-19/.configured_yyyyyyyyyyn] Error 2
make[3]: Leaving directory `/usr/src/wrt/my-feed/net/freelan'
make[2]: *** [package/feeds/Rolf/freelan/compile] Error 2
make[2]: Leaving directory `/media/rolf/DATA/tmp/compile/wrt/git'
make[1]: *** [/media/rolf/DATA/tmp/compile/wrt/git/staging_dir/target-mipsel_mips32_musl-1.1.10/stamp/.package_compile] Error 2
make[1]: Leaving directory `/media/rolf/DATA/tmp/compile/wrt/git'
make: *** [world] Error 2
```
That sounds like a simple glitch in linking which wouldn't have to be treated as an error but should be easy enough to fix.

I managed to get past that warning with "TARGET_CFLAGS += -Wno-error=cpp" now, but than ran into the following

```
/media/rolf/DATA/tmp/compile/wrt/git/staging_dir/target-mipsel_mips32_musl-1.1.10/usr/include/boost/variant/variant.hpp: In instantiation of 'void boost::variant<T0, TN>::operator!=(const U&) const [with U = asiotap::hostname_endpoint; T0_ = asiotap::hostname_endpoint; TN = {asiotap::ip_endpoint<boost::asio::ip::address_v4>, asiotap::ip_endpoint<boost::asio::ip::address_v6>}]':
build/release/libs/freelan/src/curl.cpp:149:49:   required from here
/media/rolf/DATA/tmp/compile/wrt/git/staging_dir/target-mipsel_mips32_musl-1.1.10/usr/include/boost/variant/variant.hpp:2211:9: error: static assertion failed: false && sizeof(U)
         BOOST_STATIC_ASSERT( false && sizeof(U) );
         ^
scons: *** [build/release/libs/freelan/src/curl.o] Error 1
scons: building terminated because of errors.
```

This comment isn't really OpenWRT specific but it might be related. Before I go any further I should mention that I don't have a lot of knowledge about freelan or boost and as a consequence what I say might not be correct.

Anyway, I tried to build the 2.0 release on my Archlinux machine and I get a similar error to @leggewie. 

```
g++ -o build/release/libs/freelan/src/curl.o -c -Wno-missing-field-initializers --std=c++11 -Wall -Wextra -Werror -pedantic -Wshadow -O3 -isystem third-party/install/include -DFREELAN_INSTALL_PREFIX="\"/\"" -Ibuild/release/libs/freelan/include/freelan -Ibuild/release/include build/release/libs/freelan/src/curl.cpp
build/release/libs/freelan/src/curl.cpp: In member function ‘void freelan::curl::set_proxy(const endpoint&)’:
build/release/libs/freelan/src/curl.cpp:149:13: error: could not convert ‘(& proxy)->boost::variant<T0, TN>::operator!=<asiotap::hostname_endpoint>(asiotap::hostname_endpoint::null()())’ from ‘void’ to ‘bool’
   if (proxy != asiotap::hostname_endpoint::null())
             ^
In file included from /usr/include/boost/date_time/time_duration.hpp:14:0,
                 from /usr/include/boost/date_time/posix_time/posix_time_config.hpp:16,
                 from /usr/include/boost/date_time/posix_time/posix_time_system.hpp:13,
                 from /usr/include/boost/date_time/posix_time/ptime.hpp:12,
                 from /usr/include/boost/date_time/posix_time/posix_time_types.hpp:12,
                 from /usr/include/boost/asio/time_traits.hpp:23,
                 from /usr/include/boost/asio/deadline_timer_service.hpp:27,
                 from /usr/include/boost/asio/basic_deadline_timer.hpp:25,
                 from /usr/include/boost/asio.hpp:22,
                 from build/release/libs/freelan/src/curl.hpp:51,
                 from build/release/libs/freelan/src/curl.cpp:46:
/usr/include/boost/variant/variant.hpp: In instantiation of ‘void boost::variant<T0, TN>::operator!=(const U&) const [with U = asiotap::hostname_endpoint; T0_ = asiotap::hostname_endpoint; TN = {asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6>}]’:
build/release/libs/freelan/src/curl.cpp:149:49:   required from here
/usr/include/boost/variant/variant.hpp:2211:9: error: static assertion failed: false && sizeof(U)
         BOOST_STATIC_ASSERT( false && sizeof(U) );
         ^
scons: *** [build/release/libs/freelan/src/curl.o] Error 1
```

I dug around and found out that ```proxy``` is an ```boost::variant<hostname_endpoint, ipv4_endpoint, ipv6_endpoint>```. 

Now at line 149 in ```libs/freelan/src/curl.cpp```, there is a comparison between a ```boost::variant<hostname_endpoint, ipv4_endpoint, ipv6_endpoint>``` and a ```asio::hostaname_endpoint```. Overload resolutions leads to selecting the ```boost::variant<...>::operator !=``` operator used in preventing comparison with foreign types. This, along with the other comparison operators that are part of ```boost::variant``` makes me think that comparisons can only be made between variants.

As a quick workaround I tried changing the comparison to:  ```if (proxy != asiotap::endpoint(asiotap::hostname_endpoint::null()))``` and got this:
```
build/release/libs/freelan/src/curl.cpp: In member function ‘void freelan::curl::set_proxy(const endpoint&)’:
build/release/libs/freelan/src/curl.cpp:149:13: error: ambiguous overload for ‘operator!=’ (operand types are ‘const endpoint {aka const boost::variant<asiotap::hostname_endpoint, asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6> >}’ and ‘asiotap::endpoint {aka boost::variant<asiotap::hostname_endpoint, asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6> >}’)
   if (proxy != asiotap::endpoint(asiotap::hostname_endpoint::null()))
             ^
In file included from /usr/include/boost/variant.hpp:17:0,
                 from build/release/include/asiotap/types/endpoint.hpp:51,
                 from build/release/libs/freelan/src/curl.hpp:57,
                 from build/release/libs/freelan/src/curl.cpp:46:
/usr/include/boost/variant/variant.hpp:2209:10: note: candidate: void boost::variant<T0, TN>::operator!=(const U&) const [with U = boost::variant<asiotap::hostname_endpoint, asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6> >; T0_ = asiotap::hostname_endpoint; TN = {asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6>}]
     void operator!=(const U&) const
          ^
/usr/include/boost/variant/variant.hpp:2264:17: note: candidate: bool boost::variant<T0, TN>::operator!=(const boost::variant<T0, TN>&) const [with T0_ = asiotap::hostname_endpoint; TN = {asiotap::base_ip_endpoint<boost::asio::ip::address_v4>, asiotap::base_ip_endpoint<boost::asio::ip::address_v6>}]
     inline bool operator!=(const variant& rhs) const
                 ^
In file included from build/release/libs/freelan/src/curl.hpp:57:0,
                 from build/release/libs/freelan/src/curl.cpp:46:
build/release/include/asiotap/types/endpoint.hpp:359:14: note: candidate: bool asiotap::operator!=(const endpoint&, const endpoint&)
  inline bool operator!=(const endpoint& lhs, const endpoint& rhs)
```

Which I think got us closer to selecting the operator we need (see the last two candidates). I'm not sure but I think that the operator used in preventing comparisons with foreign types is still mentioned because one variant is const and the other is not.

Since this used to work maybe the error can be related to the version of boost we're compiling against? I'm compiling against 1.58 and I see that my OpenWRT router also provides version 1.58.

--- a/libs/asiotap/include/asiotap/types/endpoint.hpp
+++ b/libs/asiotap/include/asiotap/types/endpoint.hpp
@@ -350,6 +350,9 @@ namespace asiotap
 	 */
 	std::istream& operator>>(std::istream& is, endpoint& value);
 
+// Note: this operator is defined in boost variant as of version 1.58. Keeping it around will
+// introduce overload resolution ambiguity.
+#if BOOST_VERSION < 105800
 	/**
 	 * \brief Compare two endpoints.
 	 * \param lhs The left argument.
@@ -360,6 +363,7 @@ namespace asiotap
 	{
 		return !(lhs == rhs);
 	}
+#endif
 
 	/**
 	 * \brief Get an endpoint with a default port.

--- a/libs/freelan/src/curl.cpp
+++ b/libs/freelan/src/curl.cpp
@@ -146,7 +146,7 @@ namespace freelan
 
 	void curl::set_proxy(const asiotap::endpoint& proxy)
 	{
-		if (proxy != asiotap::hostname_endpoint::null())
+		if (proxy != asiotap::endpoint(asiotap::hostname_endpoint::null()))
 		{
 			set_option(CURLOPT_PROXY, static_cast<const void*>(boost::lexical_cast<std::string>(proxy).c_str()));
 		}

